docs/contributors/coding-guidelines.md

```diff
@@ -201,6 +201,246 @@ alert( 'My name is ' + name + '.' );
+
+For additional guidance, consult the following resources:
+
+- [JSDoc Official Documentation](https://jsdoc.app/index.html)
+- [TypeScript Supported JSDoc](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc)
+
+### Custom Types
+
+Define custom types using the [JSDoc `@typedef` tag](https://jsdoc.app/tags-typedef.html).
+
+A custom type should include a description, and should always include its base type.
+
+Custom types should be named as succinctly as possible, while still retaining clarity of meaning and avoiding conflict with other global or scoped types. A `WP` prefix should be applied to all custom types. Avoid superfluous or redundant prefixes and suffixes (for example, a `Type` suffix, or `Custom` prefix). Custom types are not global by default, so a custom type does not need to be excessively specific to a particular package. However, they should be named with enough specificity to avoid ambiguity or name collisions when brought into the same scope as another type.
+
+```js
+/**
+ * A block selection object.
+ *
+ * @typedef {Object} WPBlockSelection
+ *
+ * @property {string} clientId     A block client ID.
+ * @property {string} attributeKey A block attribute key.
+ * @property {number} offset       An attribute value offset, based on the rich
+ *                                 text value.
+ */
+```
+
+Custom types can also be beneficial when describing a set of predefined options. While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids a situation where aligning tags while still respecting a maximum line length of 80 characters becomes challenging due to the space occupied by a literal values union.
+
+```js
+/**
+ * Named breakpoint sizes.
+ * 
+ * @typedef {"huge"|"wide"|"large"|"medium"|"small"|"mobile"} WPBreakpoint
+ */
+```
+
+### Importing and Exporting Types
+
+Use the [TypeScript `import` function](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#import-types) to import type declarations from other files or third-party dependencies.
+
+Since an imported type declaration can occupy an excess of the available line length and become verbose when referenced multiple times, you are encouraged to create an alias of the external type using a `@typedef` declaration at the top of the file, immediately following [the `import` groupings](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports).
+
+```js
+/** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */
```

Comment:
I think @dsifford raised this already, we should use either `'` or `"` in all places.

Reply:
> I think @dsifford raised this already, we should use either `'` or `"` in all places.

Updated in 88a12b1f3, and added additional guidelines in 4ba73de83.